satcom/csp/csp_packet.py

```diff
@@ -0,0 +1,188 @@
+        if self.destination_port < 0 or self.destination_port > 63:
+            return ValueError('PacketHeader.destination_port must be 0-63')
+        if self.source_port < 0 or self.source_port > 63:
+            return ValueError('PacketHeader.source_port must be 0-63')
+        return None
+
+    def to_bytes(self):
+        """Packs CSP packet header metadata into a bytearray"""
+        header = 0
+        cursor = 0
+        bitmask32 = 0xFFFFFFFF
+
+        cursor += FLEN_PRIO
+        header |= (self.priority << (32 - cursor)) & bitmask32
+
+        cursor += FLEN_ADDR
+        header |= (self.source << (32 - cursor)) & bitmask32
+
+        cursor += FLEN_ADDR
+        header |= (self.destination << (32 - cursor)) & bitmask32
+
+        cursor += FLEN_PORT
+        header |= (self.destination_port << (32 - cursor)) & bitmask32
+
+        cursor += FLEN_PORT
+        header |= (self.source_port << (32 - cursor)) & bitmask32
+
+        cursor += FLEN_FLAGS
+        header |= (0 << (32 - cursor)) & bitmask32
+
+        bs = bytearray(HEADER_LENGTH_BYTES)
+        bs = utils.pack_uint_big_endian(header)
+
+        return bs
+
+    @classmethod
+    def from_bytes(cls, bs: bytearray):
+        """Hydrates the CSP packet header metadata from a bytearray"""
+        if len(bs) != HEADER_LENGTH_BYTES:
+            raise ValueError('unexpected header length')
+
+        hdr = utils.unpack_uint_big_endian(bs)
+        offset = 0
+        bitmask32 = 0xFFFFFFFF
+
+        val = (hdr << offset) & bitmask32
+        cls.priority = val >> (32 - FLEN_PRIO)
+        offset += FLEN_PRIO
+
+        val = (hdr << offset) & bitmask32
+        cls.source = val >> (32 - FLEN_ADDR)
+        offset += FLEN_ADDR
+
+        val = (hdr << offset) & bitmask32
+        cls.destination = val >> (32 - FLEN_ADDR)
+        offset += FLEN_ADDR
+
+        val = (hdr << offset) & bitmask32
+        cls.destination_port = val >> (32 - FLEN_PORT)
+        offset += FLEN_PORT
+
+        val = (hdr << offset) & bitmask32
+        cls.source_port = val >> (32 - FLEN_PORT)
+        offset += FLEN_PORT
+
+        # not implemented, so ignored
+        val = (hdr << offset) & bitmask32
+        _ = val >> (32 - FLEN_FLAGS)
+        offset += FLEN_FLAGS
+
+        obj = cls(
+            priority = cls.priority,
+            source = cls.source,
+            destination = cls.destination,
+            destination_port = cls.destination_port,
+            source_port = cls.source_port
+        )
+
+        return obj
+    
+class CSPPacket():
+    def __init__(self, data: bytearray, header=None):
+        self.header = header or CSPPacketHeader()
+        self._data = data
+
+    @property
+    def data(self):
+        """Protects data from being modified after instantiation"""
+        return self._data
+
+    def err(self):
+        """Throws an error if any params are out of bounds"""
+        if self.header.err() is not None:
```

Comment:
This is a bit wasteful as the function will be called twice in the error case. We should just directly `return self.header.err()`